### PR TITLE
INFRA-820: add user-info cookie

### PR DESCRIPTION
**Pull request by jonmorehouse**
@buzzfeed/platform-infra please take a look through this first pass at adding a cookie into the `auth_proxy` to cache user-information.

This is still a WIP, I need to do some manual end to end testing and if time permits, add some unit tests in here as well. I'm open to suggestions now on how to make this better!

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -159,7 +160,7 @@ func (o *Options) Validate() error {
 
 	if len(msgs) != 0 {
 		return fmt.Errorf("Invalid configuration:\n  %s",
-			strings.Join(msgs, "\n  "))
+			strings.Join(msgs, "\n	"))
```
Comment (mccutchen) [inline options.go]: Jon I am going to forcibly remove vim from your computer

Comment (mccutchen) [inline options.go]: Also, do you know about the magic of `git add -p` which would let you skip past these lines that your vim just won't leave alone?

File: user_info_handler.go [hunk 2]
```diff
@@ -1,40 +1,122 @@
...
+		return nil, err
+	}
+
+	userInfo := make(map[string]string, 3)
+	err = json.Unmarshal([]byte(jsonString), &userInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return userInfo, errors.New("Not implemented")
```
Comment (mccutchen) [inline user_info_handler.go]: doesn't seem like the error here is accurate any more

File: options.go (hunk 1 above)
Comment (jonmorehouse) [inline options.go]: I've been trying to track down the issue ... sorry for the noise!

File: user_info_handler.go (hunk 2 above)
Comment (jonmorehouse) [inline user_info_handler.go]: yep!

File: user_info_handler.go [hunk 3]
```diff
@@ -1,40 +1,122 @@
...
+}
+
+func (m *UserInfoHandler) WriteCookie(rw http.ResponseWriter, req *http.Request, userInfo UserInfo) error {
```
Comment (jonmorehouse) [inline user_info_handler.go]: Worth mentioning that the `oauth_proxy` module has a few different methods that handle building, encrypting and setting the cookie individually. I wasn't sure that was totally necessary here so I opted to make a single method, favoring readability

Comment (jonmorehouse) [inline user_info_handler.go]: eg: https://github.com/buzzfeed/auth_proxy/blob/INFRA-820-auth-proxy-user-info-cookie/oauthproxy.go

eg: https://github.com/buzzfeed/auth_proxy/blob/14cc92a768f7e6c92a2b29095fade18391095d98/oauthproxy.go#L198

File: options.go (hunk 1 above)
Comment (yellottyellott) [inline options.go]: :joy: :joy: :joy:

File: user_info_handler.go (hunk 3 above)
Comment (mccutchen) [inline user_info_handler.go]: I assume we couldn't simply reuse those existing methods, right?

**Comment by mccutchen** [comment]
:heart: :heart: :heart: those tests! :heart: :heart: :heart: 

Looks good so far @jonmorehouse!

File: user_info_handler_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"encoding/json"
+	"github.com/bmizerany/assert"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+func TestUserInfoLoadCookiedData(t *testing.T) {
```
Comment (jonmorehouse) [inline user_info_handler_test.go]: I removed the large integration test, as I found it more useful to test each of the individual parts of the UserInfoHandler in isolation!

**Comment by jonmorehouse** [comment]
RFR @buzzfeed/platform-infra

File: main.go [hunk 5]
```diff
@@ -65,6 +65,7 @@ func main() {
 
 	flagSet.String("auth-api-url", "", "[http://]<addr>:<port>/<path> of the BuzzFeed Auth API endpoint")
 	flagSet.Duration("auth-api-refresh", time.Duration(30), "auth-api refresh")
```
Comment (mccutchen) [inline main.go]: What time frame does this actually default to?

**Comment by jonmorehouse** [comment]
@mccutchen regarding comment about using existing methods, I'm not sure we could use them! My original pass at this involved changing `oauthproxy` to use the same methods and it turned into a rabbit hole! 

I agree this is more code, but I (hope/think) its more isolated and less intrusive on the other bits of the codebase.

File: main.go (hunk 5 above)
Comment (jonmorehouse) [inline main.go]: LOL no clue!!!

Comment (jonmorehouse) [inline main.go]: Let me do some digging, I'd like for this to default to an hour

**Comment by mccutchen** [comment]
> I agree this is more code, but I (hope/think) its more isolated and less intrusive on the other bits of the codebase.

Agreed, after poking around a little bit myself

File: main.go (hunk 5 above)
Comment (mccutchen) [inline main.go]: Yeah, I think an hour at a minimum.

**Comment by jonmorehouse** [comment]
Fixed a few things, ready for a final round of review I hope!

We had an issue where `AuthApiExpire` was being set to 0 and ignoring the flag value - causing the cookie to be expired immediately. I resolved that in this latest commit.

I also tested in my VM to ensure that `auth_api/refresh` is _only_ called when the cookie is not set.

File: cookie/cookies.go [hunk 6]
```diff
@@ -29,6 +30,7 @@ func Validate(cookie *http.Cookie, seed string, expiration time.Duration) (value
 	if checkHmac(parts[2], sig) {
 		ts, err := strconv.Atoi(parts[1])
 		if err != nil {
+			log.Fatal("Missing pieces after")
```
Comment (jonmorehouse) [inline cookie/cookies.go]: Need to remove this!

File: options.go [hunk 7]
```diff
@@ -71,7 +72,7 @@ func NewOptions() *Options {
 		HttpAddress:         "127.0.0.1:4180",
 		HttpsAddress:        ":443",
 		DisplayHtpasswdForm: true,
-		CookieName:          "_oauth2_proxy",
+		CookieName:          "_auth_proxy",
```
Comment (jonmorehouse) [inline options.go]: worth pointing out that I updated this cookie to use `_auth_proxy` so there isn't a conflict with `oauth_proxy` cookies (its unlikely, but still...)

**Comment by mccutchen** [comment]
:+1: :100: :fire: 

Nice work @jonmorehouse.  Great to have some tests, and we can see about DRYing up the cookie-related code in the future.

Go ahead and :hammer: and and bump the version and we can cut a release and ship it to geodash!

File: options.go (hunk 7 above)
Comment (yellottyellott) [inline options.go]: I think there are spots in main.go and the cfg file example with this too.

**Comment by jonmorehouse** [comment]
Squashed!

File: options.go (hunk 7 above)
Comment (yellottyellott) [inline options.go]:
```
auth_proxy >> ag _oauth2_proxy                                                git:INFRA-820-auth-proxy-user-info-cookie
contrib/oauth2_proxy.cfg.example
69:# cookie_name = "_oauth2_proxy"

main.go
49:	flagSet.String("cookie-name", "_oauth2_proxy", "the name of the cookie that the oauth_proxy creates")

oauthproxy_test.go
415:	assert.Equal(t, "Cookie \"_oauth2_proxy\" not present", err.Error())

README.md
113:  -cookie-key="_oauth2_proxy": the name of the cookie that the oauth_proxy creates
```

Comment (jonmorehouse) [inline options.go]: made an update!

**Comment by yellottyellott** [comment]
I think we have some failing tests as well?

```
vagrant@vagrant-ubuntu-trusty-64:/opt/go/src/github.com/buzzfeed/auth_proxy$ ./test.sh
# github.com/buzzfeed/auth_proxy/cookie
cookie/cookies.go:12: imported and not used: "log"
FAIL	github.com/buzzfeed/auth_proxy [build failed]
ok  	github.com/buzzfeed/auth_proxy/api	0.004s
# github.com/buzzfeed/auth_proxy/cookie
cookie/cookies.go:12: imported and not used: "log"
FAIL	github.com/buzzfeed/auth_proxy/cookie [build failed]
FAIL	github.com/buzzfeed/auth_proxy/providers [build failed]
```

**Comment by jonmorehouse** [comment]
```~buzzfeed/auth_proxy (INFRA-820-auth-proxy-user-info-cookie)$ go test .
ok      _/Users/jonmorehouse/Documents/buzzfeed/auth_proxy      0.021s
```

Fixed that log import as well!

**Comment by yellottyellott** [comment]
ah, that last commit fixed the tests! :+1:

**Comment by yellottyellott** [comment]
:raised_hands: @jonmorehouse

**Comment by mccutchen** [comment]
@yellottyellott thanks for doing some actual review on this!  This ready to go?

**Comment by yellottyellott** [comment]
I think so!

**Comment by mccutchen** [comment]
:rocket:

**Comment by jonmorehouse** [comment]
:heart_eyes:

**Pull request merged**